core/common/scale.cpp

```diff
@@ -0,0 +1,493 @@
+            + 3;                // 3 == 0b11 means 4-th case of encoding
+
+        result.push_back(header);
+
+        BigInteger v{value};
+        for (size_t i = 0; i < bigIntLength; ++i) {
+          result.push_back(static_cast<uint8_t>(v % 256));
+          v >>= 8;
+        }
+
+        return expected::Value{result};
+      }
+
+      Decoder::Decoder(const std::vector<uint8_t> &source)
+          : source_{source},
+            current_iterator_{std::begin(source_)},
+            bytes_left_{static_cast<int64_t>(source_.size())} {}
+
+      std::optional<uint8_t> Decoder::nextByte() {
+        if (!hasMore()) {
+          return {};
+        }
+
+        --bytes_left_;
+        return *current_iterator_++;
+      }
+
+      std::optional<uint8_t> Decoder::previewByte() {
+        if (current_iterator_ != source_.end()) {
+          return *current_iterator_;
+        }
+        return {};
+      }
+
+      boost::optional<Decoder::IntegerResult> Decoder::decodeInteger() {
+        auto first_byte = nextByte();
+        if (!first_byte.has_value()) {
+          return {};
+        }
+
+        const uint8_t flag = first_byte.value() & 0b11;
+
+        size_t number = 0;
+
+        switch (flag) {
+          case 0b00: {
+            number = static_cast<size_t>(first_byte.value() >> 2);
+            break;
+          }
+
+          case 0b01: {
+            auto second_byte = nextByte();
+            if (!second_byte.has_value()) {
+              return {};
+            }
+
+            number = (static_cast<size_t>(first_byte.value() & 0b11111100)
+                      + static_cast<size_t>(second_byte.value()) * 256)
+                >> 2;
+            break;
+          }
+
+          case 0b10: {
+            number = first_byte.value();
+            size_t multiplier = 256;
+            if (!hasMore(3)) {
+                // not enough data to decode integer
+                return {};
+              }
+
+            for (int i = 0; i < 3; ++i) {
+                // we assured that there are 3 more bytes,
+                // no need to make checks in a loop
+              number += nextByte().value() * multiplier;
+              multiplier = multiplier << 8;
+            }
+            number = number >> 2;
+            break;
+          }
+
+          case 0b11: {
+            uint8_t m = static_cast<uint8_t>((first_byte.value() >> 2) + 4);
+            if (!hasMore(m)) {
+              // not enough data to decode integer
+              return {};
+            }
+
+            BigInteger multiplier{1};
+            BigInteger length = 0;
+            for (uint8_t i = 1; i < m; ++i) {
+                // we assured that there are m more bytes,
+                // no need to make checks in a loop
+              length += nextByte().value() * multiplier;
+              multiplier << 8;
+            }
+
+            return {{flag, length}};  // special case
+          }
+
+          default:
+            break;  // unreachable code
```

Comment:
insert assert false, just-in-case :)

Reply:
as I was told before, asserts are removed in release-mode, so use std::abort then?)
The code above can have obviousy only 4 possible cases, all of them were taken into account.
I don't think adding std::abort or assert is a good idea.